canonical-docs/phone-docs
# Replace Flask app with nginx

The phone docs application has been using *a lot* of memory (see https://private-fileshare.canonical.com/~hloeung/tmp/EWX2DYCY.png).

Since phone docs is actually just a load of static HTML files (once it's built), and all the clever redirecting that Flask does can actually be achieved with nginx, there is no need for Flask.

We already use Nginx to serve our static Jekyll-built sites (e.g. docs.snapcraft.io), so this borrows significantly from that solution. The difference is that those sites use `jekyll serve` in local development, whereas here I'm hacking the ./run script to run an `nginx` image directly.

This solution (hacking the run script) isn't ideal, but after experimenting for a while it's the best I could come up with. We can iterate on a better model down the line.

So here I've:

- Removed the Flask app (`app.py`, `routing.py`)
- Reduced `requirements.txt` to only `documentation-builder`
- Removed `entrypoint` (which ran Talisker/Gunicorn)
- Added `nginx.conf` and `redirects.map`
- Updated the Dockerfile
- Hacked `./run serve` to do `docker run nginx` (instead of `yarn run serve`)
- Removed `serve` from `package.json`, as this is now run directly in the `./run` script

QA
--

*Add a custom redirect*

Add a line like `/somewhere /somewhere/else;` to `redirects.map`. We can check this works later.

*Test the run script*

``` bash
./run clean
./run
```

Browse to http://localhost:8201, and check the phone docs work. Go to http://localhost:8201/somewhere, see that you're redirected to `/somewhere/else` (404).

*Test the Dockerfile*

``` bash
docker build --tag phone-docs --build-arg COMMIT_ID=`git rev-parse --short HEAD` .
docker run -p 80:80 phone-docs  # NB: Check nothing else is running on 80, e.g. microk8s
```

Go to http://localhost and http://localhost/somewhere, check it works.
